### PR TITLE
ASoC:SOF:fix the NOCODEC and HDA mutually incompatible

### PR DESCRIPTION
**Pull request by zhuyingjiang**
for make NOCODEC and HDA mutually incompatible, the DSP
common control part is still needed for NOCODEC to set
the DSP registers.
This fix [#1072](https://github.com/thesofproject/sof/issues/1072) [#1073](https://github.com/thesofproject/sof/issues/1073)
Signed-off-by: Zhu Yingjiang <yingjiang.zhu@linux.intel.com>

### CODE REVIEW

File: sound/soc/sof/intel/hda-dsp.c [hunk 1]
```diff
@@ -302,14 +300,12 @@ static int hda_suspend(struct snd_sof_dev *sdev, int state)
...
 	}
 
-#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
 	/* disable ppcap interrupt */
 	snd_hdac_ext_bus_ppcap_int_enable(bus, false);
 	snd_hdac_ext_bus_ppcap_enable(bus, false);
 
 	/* disable hda bus irq and i/o */
 	snd_hdac_bus_stop_chip(bus);
-#endif
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-dsp.c]: no. we don't want any dependency on the hdac library. NAK

**Comment by wenqingfu** [comment]
Making hdac library dependency optional is a feature, not a bug, for this project. @zhuyingjiang It sounds twisted but you need to find another way to fix the logic why NOCODEC depends on HDA. If needed, please open a discussion Issue for idea contribution.

**Comment by ranj063** [comment]
@zhuyingjiang we dont seem to have any issues at first boot without any dependency with the hdac library. So I am guessing we're probably missing programming certain registers coming back from S3.

**Comment by zhuyingjiang** [comment]
@plbossart There are two issues  [#1072](https://github.com/thesofproject/sof/issues/1072) and [#1073](https://github.com/thesofproject/sof/issues/1073) as described in the commit message. As I debug on Up2, they are the same root cause by commit [9a5e2e4ff4](https://github.com/thesofproject/linux/commit/9a5e2e4ff4d12e7988db52473ca86ccbafb352cb) and [7b12f064](https://github.com/thesofproject/linux/commit/7b12f06406c9a0f89417b157b829f8eb4ba04ca0). 
It means in our current code, and in NOCODEC mode, when suspend to S3, the DSP will not be powerup during resume.
> [40557.134200] sof-audio-pci 0000:00:0e.0: error: timeout on core powerup

but as @ranj063 said, and I tested on Up2, there is no problem when we only sleep to S1 by `rtcwake -s 2 -m freeze`, and there is no problem when the DSP do the first suspend/resume during the kernel startup. so still should be a PM bug. 
I guess it should be that we have done more during suspend to S3 or done less during resume from it.
In principle, we can use HOST DMA and DSP related registers in NOCODEC mode, and exclude HDA link and HDA codec related, right?
I will update the PR to do this without hda ext library.

**Comment by zhuyingjiang** [comment]
> Making hdac library dependency optional is a feature, not a bug, for this project. @zhuyingjiang It sounds twisted but you need to find another way to fix the logic why NOCODEC depends on HDA. If needed, please open a discussion Issue for idea contribution.

@wenqingfu As I talked with @RanderWang, in NOCODEC mode, we would use the HDA Host DMA and DSP control part, but should not use HDA link DMA and HDA codec part. I will update the PR.

**Comment by RanderWang** [comment]
> @zhuyingjiang we dont seem to have any issues at first boot without any dependency with the hdac library. So I am guessing we're probably missing programming certain registers coming back from S3.

yes, I agree with your idea

**Comment by zhuyingjiang** [comment]
@plbossart To make the DSP suspend/resume during the system ACPI PM flow(via `state_store` and `pm_suspend`), some DSP registers as INCTL and PPCTL is needed to be set. I re-programming-ed the `snd_hdac_ext_bus_ppcap_enable`, `snd_hdac_ext_bus_ppcap_int_enable` and `azx_int_enable` as these should be the common DSP control part for both HDA and NOCODEC.
Could you help check if this is the proper way to fix #[1702](https://github.com/thesofproject/sof/issues/1072) and #[1703](https://github.com/thesofproject/sof/issues/1073)?

File: sound/soc/sof/intel/hda-dsp.c [hunk 2]
```diff
@@ -302,14 +302,13 @@ static int hda_suspend(struct snd_sof_dev *sdev, int state)
...
 
-#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
 	/* disable ppcap interrupt */
-	snd_hdac_ext_bus_ppcap_int_enable(bus, false);
-	snd_hdac_ext_bus_ppcap_enable(bus, false);
+	hda_dsp_ctrl_ppcap_enable(sdev, false);
+	hda_dsp_ctrl_ppcap_int_enable(sdev, false);
 
-	/* disable hda bus irq and i/o */
-	snd_hdac_bus_stop_chip(bus);
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-dsp.c]: why is this removed?

File: sound/soc/sof/intel/hda-dsp.c [hunk 3]
```diff
@@ -377,18 +388,16 @@ static int hda_resume(struct snd_sof_dev *sdev)
...
 
-#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
-	/* turn off the links that were off before suspend */
-	list_for_each_entry(hlink, &bus->hlink_list, list) {
-		if (!hlink->ref_count)
-			snd_hdac_ext_bus_link_power_down(hlink);
-	}
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR, SOF_HDA_INTCTL,
+				SOF_HDA_INT_CTRL_EN | SOF_HDA_INT_GLOBAL_EN,
+				SOF_HDA_INT_CTRL_EN | SOF_HDA_INT_GLOBAL_EN);
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-dsp.c:395]: I wonder if there is a disconnect here. It's fine to use the hdac library when SOF_HDA is defined. It's not ok when it's not. You don't have to replace all the code everywhere to get rid of hdac, only where we don't want it.

Comment (zhuyingjiang) [reply sound/soc/sof/intel/hda-dsp.c:395]: @plbossart OK, changed, now should be easy to review. My first purpose it to rewrite a general part for both HDA and NOCODEC, and moved HDA part below together.

File: sound/soc/sof/intel/hda-dsp.c (hunk 2 above)
Comment (zhuyingjiang) [reply sound/soc/sof/intel/hda-dsp.c]: OK, changed, thanks!

**Comment by zhuyingjiang** [comment]
@plbossart Updated, keep the HDA part with HDAC lib.

File: sound/soc/sof/intel/hda-dsp.c [hunk 4]
```diff
@@ -377,6 +388,16 @@ static int hda_resume(struct snd_sof_dev *sdev)
 			"error: failed to ready controller during resume\n");
 		return ret;
 	}
+
```
Comment (ranj063) [inline sound/soc/sof/intel/hda-dsp.c:391]: @zhuyingjiang can you please add a comment just like in the suspend case?
